### PR TITLE
Resolver to support JSON configuration and user-pass authentication

### PR DESCRIPTION
**Pull request by mrkane27**
Summary:

Ref #664 

Modifies `Resolver` to support:

  * basic authentication
  * multiple Maven repositories at the same time
  * reading repository, Maven, and artifact settings from a file
  * optionally overridding these from the command line.

The PR a bit wobbly, as some of the `Resolver` code is now only used in the integration tests. Will fix, but wanted to make sure this is something you guys are ok having merged :D

Test Plan:

Haven't touched the unit tests yet, but tried it out on some artifacts
I needed to download.

### CODE REVIEW

**Comment by sdwilsh** [comment]
> multiple Maven repositories at the same time

Don't we already have that with this?  https://buckbuild.com/concept/buckconfig.html#maven_repositories


I guess it's a little hard to understand what this is doing because there are no documentation changes :)

**Comment by mrkane27** [comment]
I might be misunderstanding the model, but as far as I can tell:

  * `remote_file` and `buck fetch` use `[maven_repositories]`, but have no understanding of transitive dependencies;
  * `Resolver` (or `maven-importer` as aliased in `.buckconfig) is a command line tool that takes a bunch of artifacts, downloads them along with transitive deps, and sets up `prebuilt_jar` rules for all of them. It seems to be using some Eclipse Maven tooling. I don't know if there is any documentation for it to begin with. It was initially added [here](https://github.com/facebook/buck/commit/941d4deb7881dde6252c70e8ad1de112d38d86bd).

**Comment by sdwilsh** [comment]
Documentation would be fantastic to have.  People have asked about this, and I keep forgetting it exists.  You can do that in a follow-up PR if you'd like though.  Documenting it does mean your work can be used more widely. ;)

File: src/com/facebook/buck/maven/ArtifactConfig.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
+package com.facebook.buck.maven;
```
Comment (sdwilsh) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: You'll need to add the traditional copyright header to this file.

File: src/com/facebook/buck/maven/ArtifactConfig.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+import com.google.common.base.Charsets;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.annotations.SerializedName;
```
Comment (sdwilsh) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: I /think/ the Buck project is trying to standardize on Jackson, but @Coneko would know for sure.

File: src/com/facebook/buck/maven/ArtifactConfig.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.List;
+
+public class ArtifactConfig {
+
+  private static class CmdLineArgs {
+    @Argument(usage="Artifacts to resolve", metaVar="artifact")
+    public List<String> artifacts = new ArrayList<>();
+
+    @Option(name="-repo", usage="Buck repository root")
+    public String buckRepoRoot;
```
Comment (sdwilsh) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: It's not clear from this if it's meant to be Buck's root or the project's Buck root (as in where the `.buckconfig` lives)

File: src/com/facebook/buck/maven/ArtifactConfig.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    parser.parseArgument(args);
+
+    ArtifactConfig artifactConfig;
+
+    // If the -config argument was specified, load a config from JSON.
+    if (parsedArgs.artifactConfigJson != null) {
+      Gson gson = new GsonBuilder().create();
+      String jsonData = com.google.common.io.Files.toString(
+          new File(parsedArgs.artifactConfigJson), Charsets.UTF_8);
+      artifactConfig = gson.fromJson(jsonData, ArtifactConfig.class);
```
Comment (sdwilsh) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: If we're using args4j, it'd be better to just support `@` files here unless there's a lot of value in supporting JSON.

**Comment by sdwilsh** [comment]
An integration test would be nice to have too, at least using a local maven repo.

File: src/com/facebook/buck/maven/ArtifactConfig.java (hunk 2 above)
Comment (Coneko) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: I can confirm.

File: src/com/facebook/buck/maven/ArtifactConfig.java (hunk 4 above)
Comment (mrkane27) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: It's mostly to do with credentials, which I'd be _slightly_ more comfortable having in a JSON than over the command line.

Comment (sdwilsh) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: The help text doesn't make that clear to me then :)

**Comment by mrkane27** [comment]
Happy to follow up with a documentation PR.

File: src/com/facebook/buck/maven/ArtifactConfig.java (hunk 1 above)
Comment (mrkane27) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: :+1:

File: src/com/facebook/buck/maven/ArtifactConfig.java (hunk 4 above)
Comment (mrkane27) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: I've fixed it up a bit, but the format of that JSON will still have to be specified in the documentation bit. Alternatively, I can modify `usage()` to show the format as well.

**Comment by mrkane27** [comment]
Another thing I'd be interested in doing would be to give the Maven importer a mode of operation where it produces [`remote_file`](https://buckbuild.com/rule/remote_file.html) rules instead of immediately downloading everything.

**Comment by Coneko** [comment]
That sounds like a good idea, but `prebuilt_jar` targets support `maven_coords`, maybe that would work better.

I'm not sure whether those only work with `buck publish` or also with `buck fetch` though, cc @EugeneSusla .

**Comment by sdwilsh** [comment]
> That sounds like a good idea, but |prebuilt_jar| targets support
> |maven_coords|, maybe that would work better.

So, the problem is that we'd also have to toss on a sha1 there for 
hashing (and every rule that has `maven_coords`).  `maven_coords` is 
used by `buck publish` to know when to not package something and declare 
it as a dep.

File: src/com/facebook/buck/maven/ArtifactConfig.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2015-present Facebook, Inc.
```
Comment (sdwilsh) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: 2016, BTW

File: src/com/facebook/buck/maven/ArtifactConfig.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    public List<String> repositoryURIs = new ArrayList<>();
+
+    @Option(name = "-visibility", usage = "Targets that can see the artifacts")
+    public List<String> visibility = new ArrayList<>();
+
+    @Option(name = "-json", usage = "JSON configuration file for artifacts, paths and maven repos")
+    public String artifactConfigJson;
+
+    @Option(name = "-help", help = true)
+    public boolean showHelp;
```
Comment (sdwilsh) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: I missed this before; why do you have a `-` before each of the words?

File: src/com/facebook/buck/maven/ArtifactConfig.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    @Option(name = "-local-maven", usage = "Local Maven repository")
+    public String mavenLocalRepo;
+
+    @Option(name = "-maven", usage = "Maven URI(s)")
+    public List<String> repositoryURIs = new ArrayList<>();
+
+    @Option(name = "-visibility", usage = "Targets that can see the artifacts")
+    public List<String> visibility = new ArrayList<>();
+
+    @Option(name = "-json", usage = "JSON configuration file for artifacts, paths and maven repos")
```
Comment (sdwilsh) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: oxford comma

File: test/com/facebook/buck/maven/ArtifactConfigTest.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2015-present Facebook, Inc.
```
Comment (sdwilsh) [inline test/com/facebook/buck/maven/ArtifactConfigTest.java]: 2016

File: src/com/facebook/buck/maven/ArtifactConfig.java (hunk 6 above)
Comment (mrkane27) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: That's what args4j seems to [suggest](https://github.com/kohsuke/args4j/blob/master/args4j/examples/SampleMain.java).

File: src/com/facebook/buck/maven/ArtifactConfig.java (hunk 7 above)
Comment (mrkane27) [inline src/com/facebook/buck/maven/ArtifactConfig.java]: :+1:
Not sure what happened there, I love the Oxford comma.

**Comment by mrkane27** [comment]
The merge conflicts are in `BUCK.autodeps` files. I tried pulling and resolving them, however `buck autodeps` gets stuck:

    [+] PARSING BUCK FILES...384.0s

I also tried having a separate, clean, Buck checkout and running `buck` from there, same effect. Have you guys seen this before?

**Comment by sdwilsh** [comment]
@bolinfest - can we make the `buck autodeps` command handle merge 
conflicts and just blow the file away if it is found?

**Comment by mrkane27** [comment]
I also tried checking out the merge-conflicting `BUCK.autodeps` files from both my commit, and the new master - basically, valid structure, possibly incorrect contents - same effect.

**Comment by marcinkosiba** [comment]
It's possible you have a repro of the parser being stuck. It would be cool if you could help us debug:
 1) look at the buck-out/log/buck-<xx>.log coresponding to the build, see if it has any exceptions or anything interesting. Or just upload it somewhere and share it with us,
 2) take a look at cpu usage (top, atop, htop, whatever), is everything idle or is one process stuck in 100% cpu?
 3) run buck connected to the debugger and see what it's stuck in :)

**Comment by mrkane27** [comment]
screen shot 2016-03-23 at 14 47 00

Doesn't seem to be printing anything else to the log, at this point. Will try to attach the debugger now.

**Comment by mrkane27** [comment]
Here's a [gist](https://gist.github.com/mrkane27/601e3883b9fa884ce4ec) of the thread dump.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-6** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/571124226386093/int_phab) to review.

**Pull request closed without merging**